core/src/conn.rs

```diff
@@ -17,19 +17,136 @@ pub struct MongoConnection {
+        Ok(MongoConnection {
+            client,
+            current_db,
+            operation_timeout: operation_timeout.map(|to| Duration::new(to as u64, 0)),
+        })
+    }
+
+    fn get_attribute(
+        mut attributes: BTreeMap<String, String>,
+        name: &str,
+    ) -> Result<(String, BTreeMap<String, String>)> {
+        let ret = attributes.remove(name);
+        if let Some(ret) = ret {
+            Ok((ret, attributes))
+        } else {
+            let err_string = format!("uri must contain '{}' attribute", name);
+            Err(Error::UriFormatError(err_string))
+        }
     }
+
+    fn get_attributes(uri: &str) -> Result<BTreeMap<String, String>> {
+        if uri.is_empty() {
+            return Err(Error::UriFormatError("uri must not be empty".to_string()));
+        }
+        // TODO: handle DSN expansion here (i.e., lookup the DSN and add the resolved attributes).
+        // split the uri attributes on ';'
+        uri.split(';')
+            .filter(|attr| !attr.is_empty())
+            .map(|attr| {
+                // now split each attribute pair on '='
+                let mut sp = attr.split('=').map(String::from).collect::<Vec<_>>();
+                if sp.len() != 2 {
+                    return Err(Error::UriFormatError(
+                        "all uri atttributes must be of the form key=value".to_string(),
+                    ));
+                }
+                // ODBC attribute keys are case insensitive, so we lowercase the keys
+                Ok((
+                    // to_lowercase creates a String since it copies bytes
+                    sp.remove(0).trim().to_lowercase(),
+                    // trim just returns pointers into the original String/str, so we need
+                    // to_string
+                    sp.remove(0).trim().to_string(),
+                ))
+            })
+            .collect::<Result<BTreeMap<_, _>>>()
+    }
+}
+
+#[test]
+fn test_get_attributes() {
+    use crate::map;
+
+    assert!(MongoConnection::get_attributes("").is_err());
+    assert!(MongoConnection::get_attributes("Foo").is_err());
+    assert!(MongoConnection::get_attributes("driver=Foo;Bar").is_err());
+
+    let expected: BTreeMap<String, String> = map! {"driver".to_string() => "Foo".to_string()};
+    assert_eq!(
+        expected,
+        MongoConnection::get_attributes("Driver=Foo").unwrap()
+    );
+
+    let expected: BTreeMap<String, String> =
+        map! {"driver".to_string() => "Foo".to_string(), "server".to_string() => "bAr".to_string()};
+    assert_eq!(
+        expected,
+        MongoConnection::get_attributes("Driver=Foo;SERVER=bAr").unwrap()
+    );
+
+    assert_eq!(
+        expected,
+        MongoConnection::get_attributes("    Driver  =  Foo   ;    SERVER  =   bAr  ").unwrap()
+    );
+
+    assert_eq!(
+        expected,
+        MongoConnection::get_attributes("    Driver  =  Foo   ;    SERVER  =   bAr  ;").unwrap()
```

Comment:
[nit]Why using a connection string with extra blank on this test case? If I understand it correctly, this is to check that a trailing `;` is supported, right?
What about using a short connection string, like `driver=Foo;server=Bar;`? It makes the purpose of the test easier to see.

Reply:
Removing leading and trailing blanks is part of the previous test already and I don't think it's related to the trailing semicolon handling, so I thought you could remove them in this test.
You are right, I couldn't find where in the spec it says if leading and trailing blanks should be trimmed or not.
It might not because it says that `the value of the DSN keyword cannot consist only of blanks and should not contain leading blanks` and because for ADO.net this is [explicitly described](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms722656(v=vs.85)#setting-values-that-use-spaces).
But really, using leading and trailing blanks is asking for troubles.